File: erpnext/accounts/report/tds_computation_summary/tds_computation_summary.py
import frappe
from frappe import _

from erpnext.accounts.report.tax_withholding_details.tax_withholding_details import (
	TaxWithholdingDetailsReport,
)
from erpnext.accounts.utils import get_fiscal_year


class TDSComputationSummaryReport(TaxWithholdingDetailsReport):
	GROUP_BY_FIELDS = ("party_type", "party", "tax_withholding_category")
	CARRY_OVER_FIELDS = (
		"tax_id",
		"party",
		"party_type",
		"party_name",
		"tax_withholding_category",
		"party_entity_type",
		"rate",
	)
	AGGREGATE_FIELDS = ("total_amount", "tax_amount")

	def validate_filters(self):
		if self.filters.from_date > self.filters.to_date:
			frappe.throw(_("From Date must be before To Date"))

		from_year = get_fiscal_year(self.filters.from_date)[0]
		to_year = get_fiscal_year(self.filters.to_date)[0]
		if from_year != to_year:
			frappe.throw(_("From Date and To Date lie in different Fiscal Year"))

		self.filters.fiscal_year = from_year

	def get_data(self):
		return self.group_rows(super().get_data())

	def group_rows(self, data):
		grouped = {}
		for row in data:
			key = tuple(row.get(f) for f in self.GROUP_BY_FIELDS)
			bucket = grouped.setdefault(
				key,
				{
					**{f: row.get(f) for f in self.CARRY_OVER_FIELDS},
					**{f: 0.0 for f in self.AGGREGATE_FIELDS},
				},
			)

			for f in self.AGGREGATE_FIELDS:
				bucket[f] += row.get(f) or 0.0

		return list(grouped.values())

	def get_columns(self):
		party_type = self.filters.get("party_type", "Party")
		return [
			{"label": _("Tax Id"), "fieldname": "tax_id", "fieldtype": "Data", "width": 90},
			{
				"label": _(party_type),
				"fieldname": "party",
				"fieldtype": "Dynamic Link",
				"options": "party_type",
				"width": 180,
			},
			{
				"label": _(f"{party_type} Name"),
				"fieldname": "party_name",
				"fieldtype": "Data",
				"width": 180,
			},
			{
				"label": _("Tax Withholding Category"),
				"options": "Tax Withholding Category",
				"fieldname": "tax_withholding_category",
				"fieldtype": "Link",
				"width": 180,
			},
			{
				"label": _(f"{party_type} Type"),
				"fieldname": "party_entity_type",
				"fieldtype": "Data",
				"width": 180,
			},
			{"label": _("Tax Rate %"), "fieldname": "rate", "fieldtype": "Percent", "width": 120},
			{
				"label": _("Total Taxable Amount"),
				"fieldname": "total_amount",
				"fieldtype": "Float",
				"width": 120,
			},
			{"label": _("Tax Amount"), "fieldname": "tax_amount", "fieldtype": "Float", "width": 120},
		]


execute = TDSComputationSummaryReport.execute

File: erpnext/accounts/report/tds_computation_summary/tds_computation_summary.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["TDS Computation Summary"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_default('company')
		},
		{
			"fieldname":"party_type",
			"label": __("Party Type"),
			"fieldtype": "Select",
			"options": ["Supplier", "Customer"],
			"reqd": 1,
			"default": "Supplier",
			"on_change": function(){
				frappe.query_report.set_filter_value("party", "");
			}
		},
		{
			"fieldname":"party",
			"label": __("Party"),
			"fieldtype": "Dynamic Link",
			"get_options": function() {
				var party_type = frappe.query_report.get_filter_value('party_type');
				var party = frappe.query_report.get_filter_value('party');
				if(party && !party_type) {
					frappe.throw(__("Please select Party Type first"));
				}
				return party_type;
			},
			"get_query": function() {
				return {
					"filters": {
						"tax_withholding_category": ["!=",""],
					}
				}
			},
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px"
		}
	]
}
